File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        stack<int> opening;
        bool allowed[s.size()];
        string ans = "";
        memset(allowed,false,sizeof(allowed));
        
        for (int i=0;i<s.size();i++){
            auto val = s[i];
            if(val=='('){
                opening.push(i);
            }
            else if(val==')'){
                if(!opening.empty()){
                    auto temp = opening.top();
                    allowed[temp] = true;
                    allowed[i] = true;
                    opening.pop();
                }
            }
            else{
                allowed[i]= true;
            }
        }
        
        for(int i=0;i<s.size();i++){
            if(allowed[i]){
                ans+= string(1,s[i]);
            }
        }
        return ans;
    }
};

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        
        c = 0
        st = []
        for i in s:
            if i in "()":
                if i==")":
                    if c>0:
                        st.append(")")
                        c-=1
                else:
                    st.append("(")
                    c+=1
            else:
                st.append(i)
        final = []  
        while st:
            if st[-1] == "(" and c:
                c-=1
                # st.pop()
        # elif i==")":
            else:
                final.append(st[-1])
            st.pop()
        
        return "".join(reversed(final))